### DISCUSSION TITLE
Is it possible to pass messages to message queue (like RabbitMQ)?

### DISCUSSION THREAD

**Discussion by RredCat** [ThreeMammals/Ocelot#2205] [Q&A] [answered]
## Expected Behavior / New Feature
I am expecting to have a configuration that allows me to pass messages to the message queue.
It could be something like:
```
{
  "UpstreamHttpMethod": [ "GET", "POST", "PUT", "DELETE", "OPTIONS" ],
  "UpstreamPathTemplate": "/gateway/{catchAll}",
  "DownstreamPathTemplate": "/{catchAll}",
  "DownstreamScheme": "RabbitMQ",
  "DownstreamHostAndPorts": [
    { "mqOptions":"some options" }
  ]
}
```

## Actual Behavior / Motivation for New Feature
At the current moment, it is possible to use "http", "https" and "ws" as a value for "DownstreamScheme".

Is it possible to implement such behavior?

**Comment by raman-m** [comment, marked as answer]
Hello, Taras! Welcome to Ocelot world! 🐯 

> I am expecting to have a configuration that allows me to pass messages to the message queue.

Regrettably, Ocelot cannot route RabbitMQ traffic. However, if you configure upstream clients and downstream services to handle HTTP or WebSockets traffic, Ocelot should be able to route it. Downgrading the protocol to an insecure HTTP scheme is likely not in your or the community's interest. A more acceptable option would be the `wss` scheme. This reflects the present capabilities of Ocelot's pipeline. So, read this &rarr; [RabbitMQ | HTTP and WebSockets](https://www.rabbitmq.com/docs/protocols#http-and-websockets)

> At the current moment, it is possible to use "http", "https" and "ws" as a value for "DownstreamScheme".

Please read the following document: [Which protocols does RabbitMQ support?](https://www.rabbitmq.com/docs/protocols)
The documentation indicates that RabbitMQ supports several communication protocols, including [AMQP 0-9-1](https://www.rabbitmq.com/docs/protocols#amqp-091), [AMQP 1.0](https://www.rabbitmq.com/docs/protocols#amqp-10), [RabbitMQ Streams](https://www.rabbitmq.com/docs/protocols#rabbitmq-streams), [MQTT](https://www.rabbitmq.com/docs/protocols#mqtt), [STOMP](https://www.rabbitmq.com/docs/protocols#stomp), [HTTP and WebSockets](https://www.rabbitmq.com/docs/protocols#http-and-websockets)

> Is it possible to implement such behavior?

In theory, anything is possible. However, merely specifying the `RabbitMQ` scheme in the route configuration won't suffice, as more detailed protocol settings are required for a successful connection to downstream RabbitMQ endpoints. I believe the optimal solution is to use the standard AMQP protocols developed by the RabbitMQ team, which are supported by both .NET and Java libraries.
